adov_flutter/lib/widgets/date_time_picker.dart

```diff
@@ -0,0 +1,318 @@
+          keyboardAppearance: keyboardAppearance,
+          enableInteractiveSelection: enableInteractiveSelection,
+          buildCounter: buildCounter,
+        );
+      });
+
+  /// For representing the date as a string e.g.
+  /// `DateFormat("EEEE, MMMM d, yyyy 'at' h:mma")`
+  /// (Sunday, June 3, 2018 at 9:24pm)
+  final DateFormat format;
+
+  /// Called when the date chooser dialog should be shown.
+  final Future<DateTime> Function(BuildContext context, DateTime currentValue)
+  onShowPicker;
+
+  /// The [InputDecoration.suffixIcon] to show when the field has text. Tapping
+  /// the icon will clear the text field. Set this to `null` to disable that
+  /// behavior. Also, setting the suffix icon yourself will override this option.
+  final Icon resetIcon;
+
+  final TextEditingController controller;
+  final FocusNode focusNode;
+  final bool readOnly;
+  final void Function(DateTime value) onChanged;
+
+  @override
+  _DateTimeFieldState createState() => _DateTimeFieldState();
+
+  /// Returns an empty string if [DateFormat.format()] throws or [date] is null.
+  static String tryFormat(DateTime date, DateFormat format) {
+    if (date != null) {
+      try {
+        return format.format(date);
+      } catch (e) {
+        // print('Error formatting date: $e');
+      }
+    }
+    return '';
+  }
+
+  /// Returns null if [format.parse()] throws.
+  static DateTime tryParse(String string, DateFormat format) {
+    if (string?.isNotEmpty ?? false) {
+      try {
+        return format.parse(string);
+      } catch (e) {
+        // print('Error parsing date: $e');
+      }
+    }
+    return null;
+  }
+
+  /// Sets the hour and minute of a [DateTime] from a [TimeOfDay].
+  static DateTime combine(DateTime date, TimeOfDay time) => DateTime(
+      date.year, date.month, date.day, time?.hour ?? 0, time?.minute ?? 0);
+
+  static DateTime convert(TimeOfDay time) =>
+      DateTime(1, 1, 1, time?.hour ?? 0, time?.minute ?? 0);
+}
+
+class _DateTimeFieldState extends FormFieldState<DateTime> {
+  TextEditingController _controller;
+  FocusNode _focusNode;
+  bool isShowingDialog = false;
+  bool hadFocus = false;
+
+  @override
+  DateTimeField get widget => super.widget;
+
+  TextEditingController get _effectiveController =>
+      widget.controller ?? _controller;
+  FocusNode get _effectiveFocusNode => widget.focusNode ?? _focusNode;
+
+  bool get hasFocus => _effectiveFocusNode.hasFocus;
+  bool get hasText => _effectiveController.text.isNotEmpty;
+
+  @override
+  void initState() {
+    super.initState();
+    if (widget.controller == null) {
+      _controller = TextEditingController(text: format(widget.initialValue));
+      _controller.addListener(_handleControllerChanged);
+    }
+    if (widget.focusNode == null) {
+      _focusNode = FocusNode();
+      _focusNode.addListener(_handleFocusChanged);
+    }
+    widget.controller?.addListener(_handleControllerChanged);
+    widget.focusNode?.addListener(_handleFocusChanged);
```

Comment:
para que é o `?`

Reply:
em Kotlin usas o `?` para ser uma `safe call` pq se o `controller` for `null` o `addListener` não vai ser chamado, suponho que em Dart seja igual